File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/auth/StandardAuthentication.java
package uk.nhs.nhsx.core.auth;

import uk.nhs.nhsx.core.aws.secretsmanager.AwsSecretManager;
import uk.nhs.nhsx.core.aws.secretsmanager.CachingSecretManager;

import java.util.regex.Pattern;

import static uk.nhs.nhsx.core.auth.ApiKeyAuthenticator.authenticatingWithApiKey;
import static uk.nhs.nhsx.core.aws.xray.Tracing.tracing;

public class StandardAuthentication {
    private static final Pattern pattern = Pattern.compile("^[\\w_-]{6,50}");

    public static Authenticator awsAuthentication(ApiName apiName) {
        return authenticatingWithApiKey(
            apiKeyNameValidator(),
            tracing("authentication", ApiKeyAuthorizer.class,
                new CachingApiKeyAuthorizer(
                    new SecretManagerKeyAuthorizer(
                        apiName,
                        new CachingSecretManager(new AwsSecretManager())
                    )
                )
            )
        );
    }

    public static ApiKeyAuthorizer apiKeyNameValidator() {
        return key -> pattern.matcher(key.keyName).matches();
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/auth/StandardAuthentication.kt
package uk.nhs.nhsx.core.auth

import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder
import uk.nhs.nhsx.core.aws.secretsmanager.AwsSecretManager
import uk.nhs.nhsx.core.aws.secretsmanager.CachingSecretManager
import uk.nhs.nhsx.core.events.Events
import java.util.regex.Pattern

object StandardAuthentication {
    private val pattern = Pattern.compile("^[\\w_-]{6,50}")

    fun awsAuthentication(apiName: ApiName, events: Events): Authenticator {
        val cachingApiKeyAuthorizer = CachingApiKeyAuthorizer(
            SecretManagerKeyAuthorizer(
                apiName,
                CachingSecretManager(AwsSecretManager(AWSSecretsManagerClientBuilder.defaultClient())),
                events
            )
        )

        return ApiKeyAuthenticator.authenticatingWithApiKey(
            events,
            CompositeApiKeyAuthorizer(
                apiKeyNameValidator(),
                TracingApiKeyAuthorizer(cachingApiKeyAuthorizer)
            )
        )
    }

    fun apiKeyNameValidator(): ApiKeyAuthorizer = ApiKeyAuthorizer { pattern.matcher(it.keyName).matches() }
}
